src/inmanta/plugins.py

```diff
@@ -416,10 +417,20 @@ def check_requirements(self) -> None:
                 if len(result[0]) == 0:
                     raise Exception("%s requires %s to be available in $PATH" % (self.__function_name__, _bin))
 
+    @classmethod
+    def deprecate_function(cls, replaced_by: Optional[str] = None) -> None:
+        cls.deprecated = True
+        cls.replaced_by = replaced_by
+
     def __call__(self, *args: object, **kwargs: object) -> object:
         """
         The function call itself
         """
+        if Plugin.deprecated:
```

Comment:
@sanderr I noticed If I called `self.deprecated `it also worked. If I understand correctly python will first look for "deprecated" in the instance variables, and if it doesn't find it, it will look at the class variables. So I guess it is better to tell it immediately to look at the class variables or is there a reason to use `self`?

Reply:
I have never managed to make my mind up on this. I generally opt for `self` because it's polymorphic. But if (which should never be the case but still) the name is overridden on an instance, you'll get that value instead. A third option is `self.__class__.deprecated` but that's even worse imo.
Concluding, I don't think we have a strong style regarding this, so until we decide to create an ADR for it I think you can take whatever approach you prefer, though perhaps we should try to be consistent at least within each class.